alf/pretrained_models/model_adapters.py

```diff
@@ -0,0 +1,252 @@
+# limitations under the License.
+
+import types
+import copy
+from functools import partial
+from typing import Callable, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import alf
+
+
+class LoRA(nn.Module):
+    r"""Base class for LoRA (Low-Rank Adaptation).
+
+    For any model layer expressed as a matrix multiplication of the form
+    :math:`h=W_0x`, it performs a reparameterization such that:
+
+    .. math::
+
+        h = W_0x + \frac{\alpha}{r} BAx
+
+    where :math:`A\in\mathbb{R}^{r\times k}` and :math:`B\in\mathbb{R}^{d\times r}`
+    are the decomposition matrices and :math:`r` is the low-dim rank of the
+    decomposition.
+
+    The original LoRA paper doesn't adapt biases: "We leave the empirical
+    investigation of adapting the MLP layers, LayerNorm layers, and biases
+    to a future work."
+
+    If a module has a weight matrix with dimensionality greater than 2 (e.g., conv),
+    we might need to first reinterpret it as a 2D matrix ``W_0``. This reinterpretation
+    will differ for different modules.
+    """
+
+    def __init__(self,
+                 m: nn.Module,
+                 rank: int = 16,
+                 weight: float = 1.,
+                 name: str = 'LoRA'):
+        """
+        Args:
+            m: the module to be adapted
+            rank: the low rank
+            weight: weight for the low-rank weight matrix
+        """
+        super().__init__()
+
+        for para in m.parameters():
+            para.requires_grad = False
+
+        self._name = name
+        self._r = rank
+        self._alpha = weight
+        # place ``m`` in a list to avoid pytorch recursive tracking
+        self._m = [m]
+        self._merged = False
+
+        # LoRA weight mats
+        self._wA = None
+        self._wB = None  # optional; only present if m.weight is huge
+        nrows, ncols = self._decompose_weight_dims(m)
+        # original size: nrows * ncols
+        # adapter size: (nrows + ncols) * r
+        # need low-rank adaptation: (nrows + ncols) * r < nrows * ncols
+        if rank > ncols * nrows / (nrows + ncols):
+            self._wA = nn.Parameter(torch.Tensor(nrows, ncols))
+        else:
+            self._wA = nn.Parameter(torch.Tensor(rank, ncols))
+            self._wB = nn.Parameter(torch.Tensor(nrows, rank))
+        self.reset_parameters()
+        self._adapt(m)
+
+    def reset_parameters(self):
+        if self._wB is None:
+            nn.init.zeros_(self._wA)
+        else:
+            nn.init.xavier_uniform_(self._wA)
+            nn.init.zeros_(self._wB)
+
+    def _adapter_weight(self):
+        w = self._wA
+        if self._wB is not None:
+            w = self._wB @ w
+        w = w.reshape(self._m[0].weight.shape)
+        scaling = self._alpha
+        if self._wB is not None:
+            # If wB is used, wB @ wA will introduce a scale of self._r, so we
+            # need to scale the weight back.
+            scaling /= self._r
```

Comment:
It seems to me `/sqrt(self._r)` is better. Assuming each dimension of `wA @ x` are independently distributed with same variance, then the variance of each dimension of `wB @ wA @ x` scales linearly with `r`, which means its magnitude scales with `sqrt(r)`

Reply:
> It seems to me `/sqrt(self._r)` is better. Assuming each dimension of `wA @ x` are independently distributed with same variance, then the variance of each dimension of `wB @ wA @ x` scales linearly with `r`, which means its magnitude scales with `sqrt(r)`

This seems correct. The LoRA paper uses 1/r actually. I've updated the comment.